lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1975,6 +1985,71 @@ void DifferentiableActivityInfo::propagateVaried(
   }
 }
 
+void DifferentiableActivityInfo::setUsefulAndPropagateToOperands(
+    SILValue value, unsigned dependentVariableIndex) {
+  // Skip already-useful values to prevent infinite recursion.
+  if (isUseful(value, dependentVariableIndex))
+    return;
+  if (value->getType().isAddress()) {
+    propagateUsefulThroughAddress(value, dependentVariableIndex);
+    return;
+  }
+  setUseful(value, dependentVariableIndex);
+  // If the given value is a basic block argument, propagate usefulness to
+  // incoming values.
+  if (auto *bbArg = dyn_cast<SILPhiArgument>(value)) {
+    SmallVector<SILValue, 4> incomingValues;
+    bbArg->getSingleTerminatorOperands(incomingValues);
+    for (auto incomingValue : incomingValues)
+      setUsefulAndPropagateToOperands(incomingValue, dependentVariableIndex);
+    return;
+  }
+  auto *inst = value->getDefiningInstruction();
+  if (!inst)
+    return;
+  propagateUseful(inst, dependentVariableIndex);
```

Comment:
Note: `propagateUseful` cannot be inlined because it has multiple users (called multiple times in `propagateUsefulThroughAddress`.